jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java

```diff
@@ -292,6 +311,51 @@ public int read(byte[] b, int off, int len) throws IOException
         }
     }
 
+    public int read(ByteBuffer buffer) throws IOException
+    {
+        try (AutoLock lock = _contentProducer.lock())
+        {
+            // Calculate minimum request rate for DoS protection
+            _contentProducer.checkMinDataRate();
+
+            Content content = _contentProducer.nextContent();
+            if (content == null)
+                throw new IllegalStateException("read on unready input");
+            if (!content.isSpecial())
+            {
+                int read = get(content, buffer);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("read produced {} byte(s) {}", read, this);
+                if (content.isEmpty())
+                    _contentProducer.reclaim(content);
+                return read;
+            }
+
+            Throwable error = content.getError();
+            if (LOG.isDebugEnabled())
+                LOG.debug("read error={} {}", error, this);
+            if (error != null)
+            {
+                if (error instanceof IOException)
+                    throw (IOException)error;
+                throw new IOException(error);
+            }
+
+            if (content.isEof())
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("read at EOF, setting consumed EOF to true {}", this);
+                _consumedEof = true;
+                // If EOF do we need to wake for allDataRead callback?
+                if (onContentProducible())
+                    scheduleReadListenerNotification();
+                return -1;
+            }
+
+            throw new AssertionError("no data, no error and not EOF");
+        }
+    }
+
```

Comment:
The code in  read(ByteBuffer buffer) is duplicate of read(byte[] b ...)  except get(content, buffer).  But I didn't find a good way can avoid this duplication with out wrap byte[] to ByteBuffer.

Reply:
There are 2 methods now:

A=`read(ByteBuffer buffer)` and B=`read(byte[] b, int offset, int len)`.

Create another method:

C=`read(ByteBuffer buffer, byte[] b, int offset, int len)`, so that A calls C via `read(buffer, null, -1, -1)` and B calls C via `read(null, b, offset, len)`.

Inside C, the only line that changes can now be written as:
`int read = buffer == null ? get(content, b, offset, len) : get(content, byteBuffer);`

Reply:
Thanks, it's good approach.